### PR TITLE
Documentation/binderhub

### LINKED ISSUE #93: Add support for custom environment with Jupyter
**Is your feature request related to a problem? Please describe.**

No, it's a suggestion for improving the functional coverage of FADI.

**Describe the solution you'd like**

A data scientist can use Jupyter Hub for iteratively explore data sets and provide technical solutions to various problems.

In order to do so, she frequently has to change the Jupyter environment of her notebooks in order to include some specific package, to test alternative processing frameworks, etc. Typically, each project / use case can have one or many dedicated environments with daily or weekly undergoing changes.

FADI should foster such a dynamic adaptation of the data scientist's needs, by providing a way to efficiently manage extra dependencies.

For instance, a Web application could be provided for specifying, adapting or copying the environment right before instantiating Jupyter Hub. An interesting feature would be the possibility to inherit environments, and to share them among stakeholders.

**Describe alternatives you've considered**

The current recommanded way to do it is to adapt the Helm view file of the underlying Kubernetes cluster, and to restart the appropriate services. This is not really acceptable for a end user.

An alternative consists in specifying the additional dependencies in "conda install"-like commands at the beginning of the notebooks, but that makes these specifications notebooks-specific. It also implies the additional dependencies must be satisfied each time the notebook is loaded. Environment variables/secrets must be set in the notebooks, which raises securities issues. Etc, etc.

**Additional context**

Please have a look on how [Domino provides this features](https://docs.dominodatalab.com/en/3.6/reference/environments/Environment_management.html). Basically, a Docker file can be edited by the finale user for personalizing the environment. 

A nice optimization would consist in caching popular / recent / frequently used environments, in such a way running notebooks using these environments would be faster.

### PR DESCRIPTION
**Pull request by Sellto**
- Binderhub as a examples
      -> Status : Done 
      -> Just need a review of the README.md

### CODE REVIEW

File: doc/RANCHER_PROXMOX.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+Deploy FADI with rancher and proxmox
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: @Sellto Do you want that we merge doc related to Rancher also?

File: doc/RANCHER_PROXMOX.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+=============
+
+* [1. Upload ISO on Proxmox Node](#1-Upload-IS0-on-Proxmox-Node)
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: Most of the links do not work.

File: doc/RANCHER_PROXMOX.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+## 1. Upload ISO on Proxmox Node
+
+<a href="https://www.proxmox.com/" alt="OpenLDAP"> <img src="images/logos/Proxmox.png" width="150px" /></a>
+
+> "Proxmox VE is a complete open-source platform for enterprise virtualization. With the built-in web interface you can easily manage VMs and containers, software-defined storage and networking, high-availability clustering, and multiple out-of-the-box tools on a single solution."
+
+First you need to download the iso rancheros-proxmoxve-autoformat.iso, you can download by clicking [here](https://github.com/rancher/os/releases/download/v1.5.5/rancheros-proxmoxve-autoformat.iso).
+
+Once downloaded you need to upload on your proxmox node.
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: maybe, explain how?

File: doc/RANCHER_PROXMOX.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+## 2. Install Rancher
+
+We consider you have already deploy Rancher. However we give you the instructions that we have followed to deploy our Rancher server : [https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/single-node-docker/advanced/](https://rancher.com/docs/rancher/v2.x/en/installation/other-installation-methods/single-node-docker/advanced/)
+## 3. Add docker-machine driver
+
+You need to allow Rancher to control the Proxmox. We have contributed to upgrade a existing docker-machine driver to make it compatible with Rancher.
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: Add link to the driver

File: doc/RANCHER_PROXMOX.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+![Proxmoxve](images/installation/Proxmoxve.png)
+and then fill the rest of the fields like the IP of the proxmox `i.e. proxmoxHost`, the username/password `i.e. proxmoxUserName, proxmoxUserPassword `, storage of the image file `vmImageFile ` which is in our case `local:iso/rancheros-proxmoxve-autoformat.isot` and coming down to the resources you want to allocate for your node `i.e. nodevmCpuCores, vmMemory, vmStorageSize `.
+
+### Create Cluster
+
+To create your cluster:
+
+ `Cluster`  > `Add Cluster` > `Proxmoxve`
+
+you'll need to give your cluster a name, then precise the nodes in the cluster, at first start with **one master node**, you give it a name, choose of the templates created earlier for that node and then tick all 3 boxes for `etcd`, `Control Plane` and `Worker`, then choose the kubernetes version and click `create`.
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: Not clear

Comment (alexnuttinck) [reply doc/RANCHER_PROXMOX.md]: rephrase ;-)

File: examples/binderhub/deploy.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
```
Comment (alexnuttinck) [inline examples/binderhub/deploy.sh]: Maybe delete/separate the fadi installation part form this script? If we let the fadi installation, we will need to modify the genral bash script and this one.

File: doc/RANCHER_PROXMOX.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+![Proxmoxve](images/installation/Proxmoxve.png)
+and then fill the rest of the fields like the IP of the Proxmox `i.e. proxmoxHost`, the username/password `i.e. proxmoxUserName, proxmoxUserPassword `, storage of the image file `vmImageFile ` which is in our case `local:iso/rancheros-proxmoxve-autoformat.isot` and coming down to the resources you want to allocate for your node `i.e. nodevmCpuCores, vmMemory, vmStorageSize `.
+
+### Create Cluster
+
+To create your cluster:
+
+ `Cluster`  > `Add Cluster` > `Proxmoxve`
+
+you will need to give a name to your cluster, then precise the nodes in the cluster, at first start with **one master node**, you give it a name, choose of the templates created earlier for that node and then tick all 3 boxes for `etcd`, `Control Plane` and `Worker`, then choose the kubernetes version and click `create`.
```
Comment (alexnuttinck) [inline doc/RANCHER_PROXMOX.md]: Not clear, rephrase please.

File: examples/binderhub/traefik/rbac-config.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+---
```
Comment (alexnuttinck) [inline examples/binderhub/traefik/rbac-config.yaml]: Maybe point to these files directly : https://github.com/cetic/fadi/tree/master/helm/traefik ? We don't need to duplicate them.

File: examples/binderhub/README.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
+<a href="https://binderhub.readthedocs.io/en/latest/" alt="BinderHub"><img src="images/binderhub.png" width="200px"/></a>
+# What's Binderhub
+
+*The primary goal of BinderHub is creating custom computing environments that can be used by many remote users. BinderHub enables an end user to easily specify a desired computing environment from a Git repo. BinderHub then serves the custom computing environment at a URL which users can access remotely.*
+
+*BinderHub will build Docker images out of Git repositories, and then push them to a Docker registry so that JupyterHub can launch user servers based on these images*
+
+# Add Binderhub around FADI
+
+We assume that you have already a cluster deployed. If not, you can refer and follow our [Installation guide]() until the point 1.2.2
```
Comment (alexnuttinck) [inline examples/binderhub/README.md]: Add link

File: examples/binderhub/README.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+you will also need to have a valid Docker account.
```
Comment (alexnuttinck) [inline examples/binderhub/README.md]: DockerHub?

File: examples/binderhub/README.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+
+Follow this step to have FADI with Binderhub installed into your cluster:
+1. Clone this repository and go to the binderhub example folder:
+```bash
+git clone https://github.com/cetic/fadi.git fadi
+cd $pwd/fadi/examples/binderhub
+```
+
+2. open the **config.yaml** file to set your docker credentials and the the name of your project to the following inputs:
```
Comment (alexnuttinck) [inline examples/binderhub/README.md]: Not clear what to set for the image_prefix

File: examples/binderhub/README.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+git clone https://github.com/cetic/fadi.git fadi
+cd $pwd/fadi/examples/binderhub
+```
+
+2. Edit the **config.yaml** file to set your docker credentials and the the name of your project for the following inputs:
+```yaml
+config:
+    BinderHub:
+      use_registry: true
+      image_prefix: <DOCKER_ID>/<PROJECT_NAME>-
```
Comment (alexnuttinck) [inline examples/binderhub/README.md:46]: Please explain what is Docker ID, Project Name

**Review by alexnuttinck — CHANGES_REQUESTED** [review]
Hello @Sellto, See my different comments before we merge, thanks!

File: doc/RANCHER_PROXMOX.md (hunk 1 above)
Comment (Sellto) [reply doc/RANCHER_PROXMOX.md]: No, it's a mistake... This file can be deleted from this PR

File: examples/binderhub/deploy.sh (hunk 6 above)
Comment (Sellto) [reply examples/binderhub/deploy.sh]: Ok, I deleted the part concerning FADI and explain into the README that there are two scripts.

File: examples/binderhub/traefik/rbac-config.yaml (hunk 8 above)
Comment (Sellto) [reply examples/binderhub/traefik/rbac-config.yaml]: Delete this part from the binderhub example

File: examples/binderhub/README.md (hunk 9 above)
Comment (Sellto) [reply examples/binderhub/README.md]: Added

File: examples/binderhub/README.md (hunk 11 above)
Comment (Sellto) [reply examples/binderhub/README.md]: Added a little explanation...

File: examples/binderhub/README.md (hunk 12 above)
Comment (Sellto) [reply examples/binderhub/README.md:46]: added a remarks to explain the construction of a image prefix

File: examples/binderhub/README.md (hunk 10 above)
Comment (Sellto) [reply examples/binderhub/README.md]: yes ;)

**Comment by Sellto** [comment]
I think I have considered all the remarks

**Review by banzo — APPROVED** [review]

**Pull request merged**